File: swift_reference/Views/Herd/HerdView.swift
import SwiftUI
import SwiftData

struct HerdView: View {
    @Environment(\.modelContext) private var modelContext
    @Query(sort: \Animal.tagNumber) private var animals: [Animal]

    @State private var viewModel = HerdViewModel()
    @State private var showingAddAnimal = false
    @State private var showingQuickEntry = false
    @State private var showingFilters = false

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                searchAndFilterBar

                if viewModel.filteredAnimals(animals).isEmpty {
                    emptyStateView
                } else {
                    animalList
                }
            }
            .background(Color(.systemGroupedBackground))
            .navigationTitle("Herd")
            .toolbar {
                ToolbarItem(placement: .navigationBarLeading) {
                    Button {
                        showingQuickEntry = true
                    } label: {
                        Image(systemName: "bolt.fill")
                            .foregroundColor(Theme.earthGreen)
                    }
                }

                ToolbarItem(placement: .navigationBarTrailing) {
                    Button {
                        showingAddAnimal = true
                    } label: {
                        Image(systemName: "plus")
                            .foregroundColor(Theme.earthGreen)
                    }
                }
            }
            .sheet(isPresented: $showingAddAnimal) {
                NavigationStack {
                    AnimalFormView(animal: nil)
                }
            }
            .sheet(isPresented: $showingQuickEntry) {
                NavigationStack {
                    QuickEntryView()
                }
            }
            .sheet(isPresented: $showingFilters) {
                FilterSheet(viewModel: viewModel)
            }
        }
    }

    @ViewBuilder
    private var searchAndFilterBar: some View {
        VStack(spacing: 12) {
            HStack {
                HStack {
                    Image(systemName: "magnifyingglass")
                        .foregroundColor(.secondary)
                    TextField("Search by tag, breed, or notes", text: $viewModel.searchText)
                        .textFieldStyle(.plain)
                }
                .padding(12)
                .background(Color(.systemBackground))
                .cornerRadius(10)

                Button {
                    showingFilters = true
                } label: {
                    Image(systemName: viewModel.hasActiveFilters ? "line.3.horizontal.decrease.circle.fill" : "line.3.horizontal.decrease.circle")
                        .font(.title2)
                        .foregroundColor(viewModel.hasActiveFilters ? Theme.earthGreen : .secondary)
                }
            }

            HStack {
                Text("\(viewModel.filteredAnimals(animals).count) animals")
                    .font(.caption)
                    .foregroundColor(.secondary)

                Spacer()

                Menu {
                    ForEach(HerdViewModel.SortOrder.allCases, id: \.self) { order in
                        Button {
                            viewModel.sortOrder = order
                        } label: {
                            HStack {
                                Text(order.rawValue)
                                if viewModel.sortOrder == order {
                                    Image(systemName: "checkmark")
                                }
                            }
                        }
                    }
                } label: {
                    HStack(spacing: 4) {
                        Text("Sort: \(viewModel.sortOrder.rawValue)")
                            .font(.caption)
                        Image(systemName: "chevron.down")
                            .font(.caption2)
                    }
                    .foregroundColor(Theme.earthGreen)
                }
            }

            if viewModel.hasActiveFilters {
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 8) {
                        if let status = viewModel.selectedStatus {
                            FilterChip(text: status.rawValue) {
                                viewModel.selectedStatus = nil
                            }
                        }

                        if let sex = viewModel.selectedSex {
                            FilterChip(text: sex.rawValue) {
                                viewModel.selectedSex = nil
                            }
                        }

                        Button("Clear All") {
                            viewModel.clearFilters()
                        }
                        .font(.caption)
                        .foregroundColor(.secondary)
                    }
                }
            }
        }
        .padding(Theme.standardPadding)
    }

    @ViewBuilder
    private var animalList: some View {
        List {
            ForEach(viewModel.filteredAnimals(animals)) { animal in
                NavigationLink {
                    AnimalDetailView(animal: animal)
                } label: {
                    AnimalRowView(animal: animal)
                }
                .listRowInsets(EdgeInsets(top: 8, leading: 16, bottom: 8, trailing: 16))
            }
            .onDelete(perform: deleteAnimals)
        }
        .listStyle(.plain)
    }

    @ViewBuilder
    private var emptyStateView: some View {
        VStack(spacing: 16) {
            Spacer()

            Image(systemName: "pawprint")
                .font(.system(size: 64))
                .foregroundColor(.secondary.opacity(0.5))

            if viewModel.hasActiveFilters {
                Text("No animals match your filters")
                    .font(.headline)
                    .foregroundColor(.secondary)

                Button("Clear Filters") {
                    viewModel.clearFilters()
                }
                .buttonStyle(.primary)
                .frame(width: 160)
            } else {
                Text("No animals in your herd")
                    .font(.headline)
                    .foregroundColor(.secondary)

                Text("Add your first animal to get started")
                    .font(.subheadline)
                    .foregroundColor(.secondary)

                Button("Add Animal") {
                    showingAddAnimal = true
                }
                .buttonStyle(.primary)
                .frame(width: 160)
            }

            Spacer()
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    private func deleteAnimals(at offsets: IndexSet) {
        let filtered = viewModel.filteredAnimals(animals)
        for index in offsets {
            let animal = filtered[index]
            modelContext.delete(animal)
        }
    }
}

struct AnimalRowView: View {
    let animal: Animal

    var body: some View {
        HStack(spacing: 12) {
            VStack {
                Image(systemName: animal.sex.icon)
                    .font(.title2)
                    .foregroundColor(Theme.earthGreen)
            }
            .frame(width: 44, height: 44)
            .background(Theme.earthGreen.opacity(0.1))
            .cornerRadius(8)

            VStack(alignment: .leading, spacing: 4) {
                HStack {
                    Text(animal.tagNumber)
                        .font(.headline)

                    if animal.treatmentRecords.contains(where: { $0.isWithdrawalActive }) {
                        Image(systemName: "clock.fill")
                            .font(.caption)
                            .foregroundColor(.orange)
                    }
                }

                HStack(spacing: 8) {
                    Text(animal.breed.isEmpty ? "Unknown breed" : animal.breed)
                        .font(.caption)
                        .foregroundColor(.secondary)

                    Text("•")
                        .foregroundColor(.secondary)

                    Text(animal.ageDescription)
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
            }

            Spacer()

            Text(animal.status.rawValue)
                .font(.caption)
                .fontWeight(.medium)
                .padding(.horizontal, 8)
                .padding(.vertical, 4)
                .background(statusColor(for: animal.status).opacity(0.1))
                .foregroundColor(statusColor(for: animal.status))
                .cornerRadius(4)
        }
        .padding(.vertical, 4)
    }

    private func statusColor(for status: AnimalStatus) -> Color {
        switch status {
        case .active: return .green
        case .sold: return Theme.warmBrown
        case .deceased: return .gray
        case .culled: return .orange
        case .breeding: return .blue
        case .calving: return .purple
        }
    }
}

struct FilterChip: View {
    let text: String
    let onRemove: () -> Void

    var body: some View {
        HStack(spacing: 4) {
            Text(text)
                .font(.caption)

            Button(action: onRemove) {
                Image(systemName: "xmark.circle.fill")
                    .font(.caption)
            }
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 6)
        .background(Theme.earthGreen.opacity(0.1))
        .foregroundColor(Theme.earthGreen)
        .cornerRadius(16)
    }
}

struct FilterSheet: View {
    @Environment(\.dismiss) private var dismiss
    @Bindable var viewModel: HerdViewModel

    var body: some View {
        NavigationStack {
            Form {
                Section("Status") {
                    ForEach(AnimalStatus.allCases, id: \.self) { status in
                        Button {
                            if viewModel.selectedStatus == status {
                                viewModel.selectedStatus = nil
                            } else {
                                viewModel.selectedStatus = status
                            }
                        } label: {
                            HStack {
                                Text(status.rawValue)
                                    .foregroundColor(.primary)
                                Spacer()
                                if viewModel.selectedStatus == status {
                                    Image(systemName: "checkmark")
                                        .foregroundColor(Theme.earthGreen)
                                }
                            }
                        }
                        .largeTapTarget()
                    }
                }

                Section("Sex") {
                    ForEach(AnimalSex.allCases, id: \.self) { sex in
                        Button {
                            if viewModel.selectedSex == sex {
                                viewModel.selectedSex = nil
                            } else {
                                viewModel.selectedSex = sex
                            }
                        } label: {
                            HStack {
                                Image(systemName: sex.icon)
                                    .foregroundColor(Theme.earthGreen)
                                Text(sex.rawValue)
                                    .foregroundColor(.primary)
                                Spacer()
                                if viewModel.selectedSex == sex {
                                    Image(systemName: "checkmark")
                                        .foregroundColor(Theme.earthGreen)
                                }
                            }
                        }
                        .largeTapTarget()
                    }
                }

                Section {
                    Button("Clear All Filters") {
                        viewModel.clearFilters()
                    }
                    .foregroundColor(.red)
                    .largeTapTarget()
                }
            }
            .navigationTitle("Filters")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Done") {
                        dismiss()
                    }
                }
            }
        }
        .presentationDetents([.medium])
    }
}

#Preview {
    HerdView()
        .modelContainer(for: [Animal.self, CalvingRecord.self, TreatmentRecord.self], inMemory: true)
}

File: lib/views/herd/herd_view.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import '../../utils/theme.dart';
import '../../widgets/common_widgets.dart';
import '../../providers/providers.dart';
import '../../models/database.dart';

class HerdView extends ConsumerStatefulWidget {
  const HerdView({super.key});

  @override
  ConsumerState<HerdView> createState() => _HerdViewState();
}

class _HerdViewState extends ConsumerState<HerdView> {
  String _searchText = '';
  AnimalStatus? _selectedStatus;
  AnimalSex? _selectedSex;
  _SortOrder _sortOrder = _SortOrder.tagNumber;

  @override
  Widget build(BuildContext context) {
    final animalsAsync = ref.watch(animalsProvider);

    return Scaffold(
      appBar: AppBar(
        title: const Text('Herd'),
        leading: IconButton(
          icon: const Icon(Icons.flash_on, color: AppTheme.earthGreen),
          onPressed: () => context.push('/quick-entry'),
        ),
        actions: [
          IconButton(
            icon: const Icon(Icons.add, color: AppTheme.earthGreen),
            onPressed: () => context.push('/animal/new'),
          ),
        ],
      ),
      body: Column(
        children: [
          _buildSearchAndFilter(),
          Expanded(
            child: animalsAsync.when(
              data: (animals) {
                final filtered = _filterAndSort(animals);
                if (filtered.isEmpty) {
                  return _buildEmptyState(animals.isEmpty);
                }
                return _buildAnimalList(filtered);
              },
              loading: () => const Center(child: CircularProgressIndicator()),
              error: (e, _) => Center(child: Text('Error: $e')),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildSearchAndFilter() {
    return Container(
      padding: const EdgeInsets.all(AppTheme.standardPadding),
      child: Column(
        children: [
          Row(
            children: [
              Expanded(
                child: TextField(
                  decoration: InputDecoration(
                    hintText: 'Search by tag, breed, or notes',
                    prefixIcon: const Icon(Icons.search),
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(10),
                    ),
                    contentPadding: const EdgeInsets.symmetric(
                      horizontal: 16,
                      vertical: 12,
                    ),
                  ),
                  onChanged: (value) => setState(() => _searchText = value),
                ),
              ),
              const SizedBox(width: 12),
              IconButton(
                icon: Icon(
                  _hasActiveFilters
                      ? Icons.filter_alt
                      : Icons.filter_alt_outlined,
                  color:
                      _hasActiveFilters ? AppTheme.earthGreen : Colors.grey,
                ),
                onPressed: _showFilterSheet,
              ),
            ],
          ),
          const SizedBox(height: 12),
          Row(
            children: [
              Text(
                '${_filteredCount} animals',
                style: TextStyle(
                  fontSize: 12,
                  color: Colors.grey[600],
                ),
              ),
              const Spacer(),
              PopupMenuButton<_SortOrder>(
                onSelected: (order) => setState(() => _sortOrder = order),
                itemBuilder: (context) => _SortOrder.values
                    .map(
                      (order) => PopupMenuItem(
                        value: order,
                        child: Row(
                          children: [
                            Text(order.label),
                            if (_sortOrder == order)
                              const Padding(
                                padding: EdgeInsets.only(left: 8),
                                child: Icon(Icons.check, size: 18),
                              ),
                          ],
                        ),
                      ),
                    )
                    .toList(),
                child: Row(
                  children: [
                    Text(
                      'Sort: ${_sortOrder.label}',
                      style: const TextStyle(
                        fontSize: 12,
                        color: AppTheme.earthGreen,
                      ),
                    ),
                    const Icon(
                      Icons.arrow_drop_down,
                      size: 18,
                      color: AppTheme.earthGreen,
                    ),
                  ],
                ),
              ),
            ],
          ),
          if (_hasActiveFilters) ...[
            const SizedBox(height: 12),
            SingleChildScrollView(
              scrollDirection: Axis.horizontal,
              child: Row(
                children: [
                  if (_selectedStatus != null)
                    Padding(
                      padding: const EdgeInsets.only(right: 8),
                      child: ActiveFilterChip(
                        text: _selectedStatus!.label,
                        onRemove: () =>
                            setState(() => _selectedStatus = null),
                      ),
                    ),
                  if (_selectedSex != null)
                    Padding(
                      padding: const EdgeInsets.only(right: 8),
                      child: ActiveFilterChip(
                        text: _selectedSex!.label,
                        onRemove: () => setState(() => _selectedSex = null),
                      ),
                    ),
                  TextButton(
                    onPressed: _clearFilters,
                    child: const Text('Clear All'),
                  ),
                ],
              ),
            ),
          ],
        ],
      ),
    );
  }

  Widget _buildAnimalList(List<Animal> animals) {
    final withdrawalIds =
        ref.watch(withdrawalAnimalIdsProvider).valueOrNull ?? {};
    return ListView.builder(
      itemCount: animals.length,
      itemBuilder: (context, index) {
        final animal = animals[index];
        return Dismissible(
          key: Key(animal.id),
          direction: DismissDirection.endToStart,
          background: Container(
            color: Colors.red,
            alignment: Alignment.centerRight,
            padding: const EdgeInsets.only(right: 20),
            child: const Icon(Icons.delete, color: Colors.white),
          ),
          confirmDismiss: (_) async {
            return await showDialog<bool>(
              context: context,
              builder: (ctx) => AlertDialog(
                title: const Text('Delete Animal'),
                content: Text(
                    'Delete ${animal.tagNumber}? This cannot be undone.'),
                actions: [
                  TextButton(
                    onPressed: () => Navigator.pop(ctx, false),
                    child: const Text('Cancel'),
                  ),
                  TextButton(
                    onPressed: () => Navigator.pop(ctx, true),
                    style: TextButton.styleFrom(foregroundColor: Colors.red),
                    child: const Text('Delete'),
                  ),
                ],
              ),
            );
          },
          onDismissed: (_) async {
            final db = ref.read(databaseProvider);
            await (db.delete(db.animals)
                  ..where((t) => t.id.equals(animal.id)))
                .go();
          },
          child: _AnimalRow(
            animal: animal,
            hasActiveWithdrawal: withdrawalIds.contains(animal.id),
            onTap: () => context.push('/animal/${animal.id}'),
          ),
        );
      },
    );
  }

  Widget _buildEmptyState(bool isCompletelyEmpty) {
    return EmptyStateView(
      icon: Icons.pets,
      title: isCompletelyEmpty
          ? 'No animals in your herd'
          : 'No animals match your filters',
      subtitle:
          isCompletelyEmpty ? 'Add your first animal to get started' : null,
      buttonText: isCompletelyEmpty ? 'Add Animal' : 'Clear Filters',
      onButtonTap: isCompletelyEmpty
          ? () => context.push('/animal/new')
          : _clearFilters,
    );
  }

  void _showFilterSheet() {
    showModalBottomSheet(
      context: context,
      builder: (context) => _FilterSheet(
        selectedStatus: _selectedStatus,
        selectedSex: _selectedSex,
        onStatusChanged: (status) {
          setState(() => _selectedStatus = status);
          Navigator.pop(context);
        },
        onSexChanged: (sex) {
          setState(() => _selectedSex = sex);
          Navigator.pop(context);
        },
        onClear: () {
          _clearFilters();
          Navigator.pop(context);
        },
      ),
    );
  }

  void _clearFilters() {
    setState(() {
      _searchText = '';
      _selectedStatus = null;
      _selectedSex = null;
    });
  }

  bool get _hasActiveFilters =>
      _searchText.isNotEmpty ||
      _selectedStatus != null ||
      _selectedSex != null;

  int _filteredCount = 0;

  List<Animal> _filterAndSort(List<Animal> animals) {
    var filtered = animals.toList();

    // Search filter
    if (_searchText.isNotEmpty) {
      final search = _searchText.toLowerCase();
      filtered = filtered.where((a) {
        return a.tagNumber.toLowerCase().contains(search) ||
            a.breed.toLowerCase().contains(search) ||
            a.notes.toLowerCase().contains(search);
      }).toList();
    }

    // Status filter
    if (_selectedStatus != null) {
      filtered = filtered.where((a) => a.status == _selectedStatus).toList();
    }

    // Sex filter
    if (_selectedSex != null) {
      filtered = filtered.where((a) => a.sex == _selectedSex).toList();
    }

    // Sort
    switch (_sortOrder) {
      case _SortOrder.tagNumber:
        filtered.sort((a, b) => a.tagNumber.compareTo(b.tagNumber));
        break;
      case _SortOrder.dateOfBirth:
        filtered.sort((a, b) => a.dateOfBirth.compareTo(b.dateOfBirth));
        break;
      case _SortOrder.status:
        filtered.sort((a, b) => a.status.label.compareTo(b.status.label));
        break;
      case _SortOrder.recentlyAdded:
        filtered.sort((a, b) => b.createdAt.compareTo(a.createdAt));
        break;
    }

    _filteredCount = filtered.length;
    return filtered;
  }
}

enum _SortOrder {
  tagNumber('Tag #'),
  dateOfBirth('Age'),
  status('Status'),
  recentlyAdded('Recent');

  final String label;
  const _SortOrder(this.label);
}

class _AnimalRow extends StatelessWidget {
  final Animal animal;
  final bool hasActiveWithdrawal;
  final VoidCallback onTap;

  const _AnimalRow({
    required this.animal,
    required this.hasActiveWithdrawal,
    required this.onTap,
  });

  IconData get _sexIcon {
    switch (animal.sex) {
      case AnimalSex.bull:
        return Icons.circle;
      case AnimalSex.cow:
        return Icons.circle_outlined;
      case AnimalSex.heifer:
        return Icons.brightness_1;
      case AnimalSex.steer:
        return Icons.square;
      case AnimalSex.calf:
        return Icons.square_outlined;
    }
  }

  Color get _statusColor {
    switch (animal.status) {
      case AnimalStatus.active:
        return Colors.green;
      case AnimalStatus.sold:
        return AppTheme.warmBrown;
      case AnimalStatus.deceased:
      case AnimalStatus.culled:
        return Colors.grey;
      case AnimalStatus.breeding:
        return Colors.blue;
      case AnimalStatus.calving:
        return Colors.purple;
    }
  }

  String get _ageDescription {
    final months = DateTime.now().difference(animal.dateOfBirth).inDays ~/ 30;
    if (months < 12) {
      return '$months mo';
    } else {
      final years = months ~/ 12;
      final remainingMonths = months % 12;
      if (remainingMonths == 0) {
        return '$years yr';
      }
      return '$years yr $remainingMonths mo';
    }
  }

  @override
  Widget build(BuildContext context) {
    return InkWell(
      onTap: onTap,
      child: Container(
        decoration: BoxDecoration(
          border: Border(
            bottom: BorderSide(color: Colors.grey.shade200),
          ),
        ),
        padding: const EdgeInsets.symmetric(
          horizontal: AppTheme.standardPadding,
          vertical: 12,
        ),
        child: Row(
          children: [
            Container(
              width: 44,
              height: 44,
              decoration: BoxDecoration(
                color: AppTheme.earthGreen.withValues(alpha: 0.1),
                borderRadius: BorderRadius.circular(8),
              ),
              child: Icon(_sexIcon, color: AppTheme.earthGreen),
            ),
            const SizedBox(width: 12),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    children: [
                      Text(
                        animal.tagNumber,
                        style: const TextStyle(
                          fontWeight: FontWeight.bold,
                          fontSize: 16,
                        ),
                      ),
                      if (hasActiveWithdrawal) ...[
                        const SizedBox(width: 6),
                        const Icon(
                          Icons.access_time,
                          size: 14,
                          color: Colors.orange,
                        ),
                      ],
                    ],
                  ),
                  const SizedBox(height: 2),
                  Text(
                    '${animal.breed.isEmpty ? 'Unknown breed' : animal.breed} • $_ageDescription',
                    style: TextStyle(fontSize: 13, color: Colors.grey[600]),
                  ),
                ],
              ),
            ),
            StatusBadge(
              text: animal.status.label,
              color: _statusColor,
            ),
            const SizedBox(width: 4),
            const Icon(Icons.chevron_right, color: Colors.grey, size: 20),
          ],
        ),
      ),
    );
  }
}

class _FilterSheet extends StatelessWidget {
  final AnimalStatus? selectedStatus;
  final AnimalSex? selectedSex;
  final ValueChanged<AnimalStatus?> onStatusChanged;
  final ValueChanged<AnimalSex?> onSexChanged;
  final VoidCallback onClear;

  const _FilterSheet({
    required this.selectedStatus,
    required this.selectedSex,
    required this.onStatusChanged,
    required this.onSexChanged,
    required this.onClear,
  });

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.all(AppTheme.standardPadding),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              const Text(
                'Filters',
                style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
              ),
              TextButton(
                onPressed: () => Navigator.pop(context),
                child: const Text('Done'),
              ),
            ],
          ),
          const SizedBox(height: 16),
          const Text('Status', style: TextStyle(fontWeight: FontWeight.w600)),
          const SizedBox(height: 8),
          Wrap(
            spacing: 8,
            children: AnimalStatus.values.map((status) {
              final isSelected = selectedStatus == status;
              return ChoiceChip(
                label: Text(status.label),
                selected: isSelected,
                onSelected: (selected) {
                  onStatusChanged(selected ? status : null);
                },
              );
            }).toList(),
          ),
          const SizedBox(height: 16),
          const Text('Sex', style: TextStyle(fontWeight: FontWeight.w600)),
          const SizedBox(height: 8),
          Wrap(
            spacing: 8,
            children: AnimalSex.values.map((sex) {
              final isSelected = selectedSex == sex;
              return ChoiceChip(
                label: Text(sex.label),
                selected: isSelected,
                onSelected: (selected) {
                  onSexChanged(selected ? sex : null);
                },
              );
            }).toList(),
          ),
          const SizedBox(height: 16),
          SizedBox(
            width: double.infinity,
            child: TextButton(
              onPressed: onClear,
              child: const Text(
                'Clear All Filters',
                style: TextStyle(color: Colors.red),
              ),
            ),
          ),
          SizedBox(height: MediaQuery.of(context).padding.bottom),
        ],
      ),
    );
  }
}
